// xgplayer-subtitles/es/observer.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ResizeObserver = function () {
  function ResizeObserver() {
    var _this = this;

    _classCallCheck(this, ResizeObserver);

    this.__handlers = [];
    if (!window.ResizeObserver) {
      return;
    }
    this.observer = new window.ResizeObserver(function (entries) {
      // const t = new Date().getTime()
      _this.__trigger(entries);
    });
  }

  _createClass(ResizeObserver, [{
    key: "addObserver",
    value: function addObserver(target, handler) {
      if (!this.observer) {
        return;
      }
      this.observer && this.observer.observe(target);
      var __handlers = this.__handlers;

      var index = -1;
      for (var i = 0; i < __handlers.length; i++) {
        if (__handlers[i] && target === __handlers[i].target) {
          index = i;
        }
      }
      if (index > -1) {
        this.__handlers[index].handler.push(handler);
      } else {
        this.__handlers.push({
          target: target,
          handler: [handler]
        });
      }
    }
  }, {
    key: "unObserver",
    value: function unObserver(target) {
      var i = -1;
      this.__handlers.map(function (item, index) {
        if (target === item.target) {
          i = index;
        }
      });
      this.observer && this.observer.unobserve(target);
      i > -1 && this.__handlers.splice(i, 1);
    }
  }, {
    key: "destroyObserver",
    value: function destroyObserver() {
      this.observer && this.observer.disconnect();
      this.observer = null;
      this.__handlers = null;
    }
  }, {
    key: "__runHandler",
    value: function __runHandler(target, cr) {
      var __handlers = this.__handlers;

      for (var i = 0; i < __handlers.length; i++) {
        if (__handlers[i] && target === __handlers[i].target) {
          __handlers[i].handler && __handlers[i].handler.map(function (handler) {
            try {
              handler(target, cr);
            } catch (error) {
              console.error(error);
            }
          });
          break;
        }
      }
    }
  }, {
    key: "__trigger",
    value: function __trigger(entries) {
      var _this2 = this;

      entries.map(function (item) {
        var cr = item.contentRect;
        _this2.__runHandler(item.target, cr);
      });
    }
  }]);

  return ResizeObserver;
}();

var resizeObserver = null;

function addObserver(target, handler) {
  if (!resizeObserver) {
    resizeObserver = new ResizeObserver();
  }
  resizeObserver.addObserver(target, handler);
}

function unObserver(target, handler) {
  if (!resizeObserver) {
    return;
  }
  resizeObserver.unObserver(target, handler);
}

function destroyObserver(target, handler) {
  if (!resizeObserver) {
    return;
  }
  resizeObserver.destroyObserver(target, handler);
  resizeObserver = null;
}

export { addObserver, unObserver, destroyObserver };
